### PR TITLE
fix: adding unit test suite for fullstory

### PR DESCRIPTION
**Pull request by shrouti1507**
## PR Description

Please include a summary of the change along with the relevant motivation and context.

## Linear task (optional)

resolves INT-895

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **Tests**
  - Enhanced testing capabilities for the Fullstory integration.
  - Improved initialization, event handling, and user identification tests.
  - Added new tests for pageview, track, and user identification events.

### CODE REVIEW

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 1]
```diff
@@ -1,173 +1,260 @@
...
   scriptElement.type = 'text/javascript';
   scriptElement.id = 'dummyScript';
   const headElements = document.getElementsByTagName('head');
   headElements[0].insertBefore(scriptElement, headElements[0].firstChild);
+
+  global.logger = {
+    info: jest.fn(),
+    error: jest.fn(),
+    setLogLevel: jest.fn(),
+  };
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:26]: removed

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 2]
```diff
@@ -1,173 +1,260 @@
...
+
+  delete global.window; // Ensure we start clean on each test
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:28]: if I do delete it, it is not being able to mock events properly.

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 3]
```diff
@@ -1,173 +1,260 @@
...
+    });
+
+    test('init call of FullStory', () => {
+      fullstory.init();
+      expect(typeof window.FS).toBe('object');
+    });
+
+    test('should call loadNativeSdk with correct parameters', () => {
+      fullstory.init();
+      expect(loadNativeSdk).toHaveBeenCalledWith(fullstory.fs_debug_mode, fullstory.fs_host, fullstory.fs_org);
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:90]: added

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 4]
```diff
@@ -1,173 +1,252 @@
...
+      shutdown: jest.fn(),
+      setUserVars: jest.fn(),
+      log: jest.fn(),
+      init: jest.fn(),
+    },
+    _fs_namespace: 'FS',
+    _fs_identity: jest.fn().mockImplementation(() => {
+      console.log("Mock _fs_identity called");
+    }),
+  };
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:51]: Tried but it is failing to initialize properly. Getting error :
 ● ```
FullStory › init › should initialize the destination

    expect(received).toBe(expected) // Object.is equality

    Expected: "object"
    Received: "undefined"

      83 |     test('should initialize the destination', () => {
      84 |       fullstory.init();
    > 85 |       expect(typeof window.FS).toBe('object');
         |                                ^
      86 |     });
      87 |
      88 |     test('should call loadNativeSdk with correct parameters', () => {

      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:85:32)

  ● FullStory › init › should handle cross-domain support correctly

    TypeError: fs(...) is not a function

      128 |           }
      129 |         }
    > 130 |         fs('shutdown')();
          |                       ^
      131 |         waitUntil(window._fs_identity, identify, timeout, fs('restart'));
      132 |       })();
      133 |     }

      at src/integrations/Fullstory/browser.js:130:23
      at Fullstory.init (src/integrations/Fullstory/browser.js:132:9)
      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:102:17)

  ● FullStory › page › should send pageview event

    TypeError: Cannot read properties of undefined (reading 'event')

      151 |     };
      152 |
    > 153 |     window.FS.event('Viewed a Page', Fullstory.getFSProperties(props));
          |               ^
      154 |   }
      155 |
      156 |   identify(rudderElement) {

      at Fullstory.page (src/integrations/Fullstory/browser.js:153:15)
      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:172:17)

  ● FullStory › Track › should fire track event

    TypeError: Cannot read properties of undefined (reading 'event')

      166 |
      167 |   track(rudderElement) {
    > 168 |     window.FS.event(
          |               ^
      169 |       rudderElement.message.event,
      170 |       Fullstory.getFSProperties(rudderElement.message.properties),
      171 |     );

      at Fullstory.track (src/integrations/Fullstory/browser.js:168:15)
      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:218:17)

  ● FullStory › identify › should send identify call

    TypeError: Cannot set properties of undefined (setting 'identify')

      233 |       fullstory = new Fullstory(destinationConfig, analyticsInstance, destinationInfo);
      234 |       fullstory.analytics.loadOnlyIntegrations = {};
    > 235 |       window.FS.identify = jest.fn();
          |                         ^
      236 |     });
      237 |     test('should send identify call', () => {
      238 |       const processedTraits = Fullstory.getFSProperties({

      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:235:25)

Test Suites: 1 failed, 1 total
Tests:       5 failed, 7 passed, 12 total
Snapshots:   0 total
Time:        0.469 s, estimated 1 s
```

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 5]
```diff
@@ -1,173 +1,252 @@
...
-          position: 1,
-          value: 6.0,
-          typeOfProduct: 'Food',
-          url: 'https://www.example.com/product/bacon-jam',
-          image_url: 'https://www.example.com/product/bacon-jam.jpg',
+          customProp: 'testProp',
+          checkoutId: '123456',
         },
-      ],
+      ]);
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js]: done

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 6]
```diff
@@ -1,173 +1,252 @@
...
-    expect(window.FS.identify.mock.calls[0][0]).toEqual('rudder01');
-    expect(window.FS.identify.mock.calls[0][1]).toEqual({
-      email: 'abc@ruddertack.com',
-    });
+      });
+      expect(window.FS.identify.mock.calls[0][0]).toEqual('rudder01');
+      expect(window.FS.identify.mock.calls[0][1]).toEqual({
+        email: 'abc@ruddertack.com',
+      });
+    })
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js]: done

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 7]
```diff
@@ -1,173 +1,260 @@
...
-          position: 1,
-          value: 6.0,
-          typeOfProduct: 'Food',
-          url: 'https://www.example.com/product/bacon-jam',
-          image_url: 'https://www.example.com/product/bacon-jam.jpg',
+          customProp: 'testProp',
+          checkoutId: '123456',
         },
-      ],
+      ]);
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js]: this change is done

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:51]: tried, not working.

[the remark this answers is not in the document]
Comment (shrouti1507) [reply packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:51]: even this is erroring out with :
FullStory › init › should handle cross-domain support correctly

    TypeError: fs(...) is not a function

      128 |           }
      129 |         }
    > 130 |         fs('shutdown')();
          |                       ^
      131 |         waitUntil(window._fs_identity, identify, timeout, fs('restart'));
      132 |       })();
      133 |     }

      at src/integrations/Fullstory/browser.js:130:23
      at Fullstory.init (src/integrations/Fullstory/browser.js:132:9)
      at Object.<anonymous> (__tests__/integrations/Fullstory/browser.test.js:110:17)

File: packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js [hunk 8]
```diff
@@ -1,173 +1,255 @@
...
-            email: 'abc@ruddertack.com',
+
+  describe('identify', () => {
+    let fullstory;
+    beforeEach(() => {
+      fullstory = new Fullstory(destinationConfig, analyticsInstance, destinationInfo);
+      fullstory.analytics.loadOnlyIntegrations = {};
+      window.FS.identify = jest.fn();
+    });
+    test('should send identify call', () => {
```
Comment (saikumarrs) [inline packages/analytics-js-integrations/__tests__/integrations/Fullstory/browser.test.js:237]: We should also test another case where only anonymous ID is present.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

**Pull request closed without merging**